torch/distributed/_sharded_tensor/ops/linear.py

```diff
@@ -183,27 +199,40 @@ def _handle_row_wise_sharding(input, world_size, weight, rank, local_shard_t, bi
         for idx, placement in enumerate(weight._sharding_spec.placements):
             split_size = input_split_sizes[placement.rank()]
             offset_start_idx = idx * sharded_dim_size_max
-            indices[placement.rank()] = list(range(offset_start_idx, offset_start_idx + split_size))
+            indices[placement.rank()] = list(
+                range(offset_start_idx, offset_start_idx + split_size)
+            )
         indices_flatten = list(idx for indice in indices for idx in indice)
 
-        input_t = input_t.index_select(0, torch.tensor(indices_flatten, device=input_t.device))
+        input_t = input_t.index_select(
+            0, torch.tensor(indices_flatten, device=input_t.device)
+        )
 
-    gathered_input = torch.empty(input_split_sizes[rank] * world_size, input_t_size[1], device=input_t.device)
+    output_tensor_list = [
+        torch.empty(
+            input_split_sizes[rank],
+            input_t_size[1],
+            device=input.device,
+            dtype=local_shard_t.dtype,
+            requires_grad=local_shard_t.requires_grad,
+        )
+        for _ in range(world_size)
+    ]
 
     # Perform alltoall
-    dist.all_to_all_single(gathered_input, input_t, input_split_sizes=input_split_sizes, group=pg)
-    gathered_input = gathered_input.t()
+    all_to_all(
+        output_tensor_list,
+        list(torch.split(input_t, input_split_sizes)),
```

Comment:
Seems like we have extra copies here since we are not using something like `all_to_all_single` autograd enabled function? If so, maybe we can create a gh issue for this as future improvement and add a TODO here pointing to that gh issue.

Reply:
I will implement the `all_to_all_single` then.